[GSOC2022]REFACTOR---Normalize pdclient to client-java (#2491)
pingcap/tispark: docs/design/2022-08-11-normalize-client.md

# TiSpark Design Documents

- Author(s): [`ajian`](http://github.com/ajian2002)
- Tracking Issue:[#2232](https://github.com/pingcap/tispark/issues/2232)
  ,[tikv/client-java#514](https://github.com/tikv/client-java/issues/514)

## Table of Contents

* [Introduction](#introduction)
* [Motivation or Background](#motivation-or-background)
* [Detailed Design](#detailed-design)

>     * [Build and Dependences](#Build and Dependences)

Reviewer: broken links

        * [TiConfiguration](#tiConfiguration)
        * [TIDBOptions](#tidboptions)
    * [ClientSession](#clientsession)
        * [Previous module or method call relationship](#previous-module-or-method-call-relationship)
        * [Current  module or method call relationship](#current-module-or-method-call-relationship)
* [Compatibility](#compatibility)
* [Test Design](#test-design)
    * [Unit tests](#unit-tests)
    * [Integration tests](#integration-tests)
* [Impacts & Risks](#impacts--risks)
* [Investigation & Alternatives](#investigation--alternatives)
* [References](#references)

## Introduction

Replace `tispark's tikv-client` module, use `tikv's client-java` to interact
with [`tikv`](https:///github.com/tikv/tikv).

## Motivation or Background

Until the latest version 3.0.1, the interaction between `tispark` and `tikv` still uses the built-in
client (`tikv-client` module). The upstream (`tikv/client-java`) provides a dedicated Java client to
interact with `tikv`. Due to the duplication of a large number of functions, `tispark` hopes to use
the upstream client to handle the interaction with `tikv`, and migrate its own client module related

> functions to the upstream, only Maintain some information that `tispark` itself needs to retain to

Reviewer 2: ```suggestion
functions to the upstream, only maintain some information that `tispark` itself needs to retain to
```

reduce the complexity of `tispark`.

## Detailed Design

### Build and Dependences

> First, I will import the upstream (referring to `tikv/client-java`) maven dependency.

Reviewer 2: Please give a link to `tikv/client-java` maven repo.

1. Considering that the current `tikv-client` module (referring to the internal implementation of

>    the current `tispark client`) relies on three `protobuf` repositories (`kvproto`,`raft-rs`

Reviewer 2: What is` tispark client`? Do you mean `tispark/tikv-client`?

   and `tipb`, see `tikv-client/scripts/proto.sh` for details), in addition to the need for your own

>    The maintained `tipb`, the other two can be handed over to upstream maintenance.

Reviewer 2: ```suggestion
maintained `tipb`, the other two can be handed over to upstream maintenance.
```

   package names to prevent conflicts.

### Configuration

Second, consider that there are differences and a lot of similarities between the upstream
configuration item (`TiConfiguration`) and the current configuration. So keep the current
configuration class, add the `ConverterUpstreamUtils.java` tool class, and convert the current
configuration into the upstream configuration as much as possible. The following is a partial
configuration comparison table. (See also the comments for the `convertTiConfiguration` method in
the `ConverterUpstreamUtils` class). (Not perfect, because some upstream configurations cannot be
modified).

Here are some notable configuration items

#### TiConfiguration

| item                       | tispark      | client-java | description                                     |
|----------------------------|--------------|-------------|-------------------------------------------------|
| timeout                    | 10 minutes   | 200ms       | convert                                         |
| maxFrameSize               | 2GB          | 512MB       | convert                                         |

> | netWorkMappingName         | ""           | ""          | can't be convert,but has same default value     |

Reviewer 2: ```suggestion
| netWorkMappingName         | ""           | ""          | can't be converted,but has a same default value     |
```

| ~~writeBufferSize~~        (delete) | client-java does not support setting |
| ~~writeThreadPerTask~~     (delete) | client-java does not support setting |
| ~~retryCommitSecondaryKey~~(delete) | client-java does not support setting |
| ~~prewriteMaxRetryTimes~~  (delete) | client-java does not support setting |

### ClientSession

The code of the core module takes the `TiSession` class of the client module as the entry, calls
the `createSnapshot`, `getOrCreateSnapShotCatalog`and `getPdclient` methods, and uses the snapshot,
catalog, and `pdclient` classes as the cut points to connect other codes of the entire client
module. So the key is to replace the `tiSession` class and rewrite these methods, so that the core
module can not only enjoy the upstream functions, but also flexibly rewrite and expand the upstream.
Therefore, the `ClientSession` class was added, the `TiSession`, `PDClient` and other classes were
deleted, the Catalog and Snapshot classes were retained, and the `ClientSession` was used as a new
entry for the core module. `ClientSession` holds the upstream `TiSession` and the current original

> Catalog and Snapshot.

Reviewer 3: Catalog and SnapshotCatalog? I think snapshot is not held by ClientSession?

A
end
subgraph <font color='green'>other</font>
B
end
subgraph <font color='green'>tikv-client</font>
C-->D
C-->E
C-->F
C-->H
C-->G
end
A-->C
B-->C

```

#### Current module or method call relationship

```mermaid

flowchart TD
A[core]
B[other module]
subgraph <font color='green'>core</font>
A
end
subgraph <font color='green'>other</font>
B
end
A-->C
B-->C
C[ClientSession]
D[Snapshot]
E[Catalog]
F[TiSession]
L[PDClient]
G[...]
H[tispark/TIConfiguration]
I[tikv/TIConfiguration]
C-->|<font color='red'>ticonfiguration converent</font> |F
subgraph <font color='green'>tispark/tikv-client</font>
C-->D
C-->E
C-->H
end
subgraph  <font color='green'>tikv/client-java</font>
F-->L
F-->I
F-->G
end
```

## Compatibility

- Not forward compatible, but users will not directly use `tispark client` related modules, so it
  does not affect users' use.
- When using `TiSession` afterwards, you need to use `ClientSession` instead. The method provided
  by `ClientSession` is preferred, and if it is not provided, `getTiKVSession()` is required to
  indirectly call the upstream richer method.
- It is recommended to add the `org.tikv.shade` prefix when you need to use the three

>   packages (`io.grpc`, `com.google`, `io.netty`).

Reviewer 3: We need to specify the configs which are deleted

D[Snapshot]
E[Catalog]
F[PDClient]
G[....]
H[TiConfiguaration]
subgraph <font color='green'>core</font>
A
end
subgraph <font color='green'>other</font>
B
end
subgraph <font color='green'>tikv-client</font>
C-->D
C-->E
C-->F
C-->H
C-->G
end
A-->C
B-->C

```

#### Current module or method call relationship

```mermaid

flowchart TD
A[core]
B[other module]
subgraph <font color='green'>core</font>
A
end
subgraph <font color='green'>other</font>
B
end
A-->C
B-->C
C[ClientSession]
D[Snapshot]
E[Catalog]
F[TiSession]
L[PDClient]
G[...]
H[tispark/TIConfiguration]
I[tikv/TIConfiguration]
C-->|<font color='red'>ticonfiguration converent</font> |F
subgraph <font color='green'>tispark/tikv-client</font>
C-->D
C-->E
C-->H
end
subgraph  <font color='green'>tikv/client-java</font>
F-->L
F-->I
F-->G
end
```

## Compatibility

- Not forward compatible, but users will not directly use `tispark client` related modules, so it
  does not affect users' use.
- When using `TiSession` afterwards, you need to use `ClientSession` instead. The method provided
  by `ClientSession` is preferred, and if it is not provided, `getTiKVSession()` is required to
  indirectly call the upstream richer method.

> - It is recommended to add the `org.tikv.shade` prefix when you need to use the three

Reviewer 3: it is recommended or required? if not added prefix will it cause errors? besides, you introduce how to do, it is better to introduce why in design

Author: This is the recommended practice. It may be compatible if you don't do this. No guarantees are given that it will work
